Login enhancement: Extract plugin setup web view model from view controller

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #7236 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
For #7236, we want to add a new screen for the error when a user tries to log in to a site that doesn't have WooCommerce. For this screen, we want to add a button to install Woo on a web view. This will be very similar to the flow with installing Jetpack, so this PR aims to extract the view model for Jetpack setup from the web view controller so that we can reuse the view controller logic later with Woo setup.

This PR also creates an extension of `WKWebView` and extracts some logic from `AuthenticatedWebView` to authenticate WPCom automatically and avoid forcing users to log in twice. I haven't removed the logic from `AuthenticatedWebView` though since that will require regression test for shipping labels, so I'll create a separate task for the migration after this PR is merged.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
We need to make sure that the Jetpack setup flow still works as before:
- Prerequisite: make sure to have a test site with WooCommerce and no Jetpack.
- To make sure any cookies are cleared from the app's browser, delete the app first.
- Select Enter site address.
- Proceed to log in with either WP.com credentials. After the login succeeds, notice that you reach the no Jetpack error screen.
- Select Install Jetpack and follow the instruction on the web view. 
- When you reach the Jetpack Authorization step, notice that you no longer have to enter WPCom credentials in the web view.
- After the authorization step, notice that the web view is dismissed, and the app navigates to the home screen.

### Screenshot

https://user-images.githubusercontent.com/5533851/182359228-84f5fbc6-62a2-4923-83af-799f648ee54f.mp4



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

itsmeichigo committed:
Extract view model for Jetpack setup web view

itsmeichigo committed:
Add unit test for jetpack setup web view model

itsmeichigo committed:
Rename method handleDismissal

itsmeichigo committed:
Authenticate WPCom when calling the plugin setup web view

itsmeichigo committed:
Add new extension for WKWebView to authenticate WPCom

itsmeichigo committed:
Remove sandbox env configuration

itsmeichigo committed:
Only authenticate if there are WPCom credentials

jaclync commented on WooCommerce/Classes/Authentication/Navigation Exceptions/JetpackSetupWebViewModel.swift:
```diff
@@ -13,127 +12,54 @@ final class JetpackSetupWebViewController: UIViewController {
-        ])
-    }
+    // MARK: - `PluginSetupWebViewModel` conformance
+    var title: String { Localization.title }
```
super nit: this can also be a `let`
```suggestion
    let title = Localization.title
```

itsmeichigo replied:
Fixed in 1899b57cf9a728acce705d2eb605f2e175a09079 👍

jaclync commented on WooCommerce/Classes/Authentication/Navigation Exceptions/PluginSetupWebViewController.swift:
```diff
@@ -0,0 +1,111 @@
+        guard let url = viewModel.initialURL else {
+            return
+        }
+        progressSubscription = webView.publisher(for: \.estimatedProgress)
```
this is cool!

jaclync commented on WooCommerce/Classes/Authentication/Navigation Exceptions/PluginSetupWebViewController.swift:
```diff
@@ -0,0 +1,111 @@
+
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+        if isMovingFromParent {
```
nit: we also have an extension helper `isBeingDismissedInAnyWay` to support dismissal when the view is presented in different ways:

https://github.com/woocommerce/woocommerce-ios/blob/931bb29fabacdb9f98a61989b80b2a7fa79bd8ae/WooCommerce/Classes/Extensions/UIViewController%2BNavigation.swift#L4-L8

itsmeichigo replied:
Thanks for letting me know about this! Updated in 953991735ca67ef7ccc1e8ec4b669f3abc4da11e

jaclync commented on WooCommerce/Classes/Authentication/Navigation Exceptions/PluginSetupWebViewModel.swift:
```diff
@@ -0,0 +1,18 @@
+    func handleDismissal()
+
+    /// Handler for a navigation URL
+    func decidePolicy(for navigationURL: URL, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void)
```
nit: maybe we can try using the async/await version for `WKNavigationDelegate.webView(_:decidePolicyFor:) async` as in `PluginSetupWebViewController`? 🙂

itsmeichigo replied:
Good point 👍 Switched to use the new version in 45e46f67ba3ccfddaa50f0ec4fd4b5b3ee7bf5e4.

jaclync commented on WooCommerce/Classes/Authentication/Navigation Exceptions/PluginSetupWebViewController.swift:
```diff
@@ -0,0 +1,111 @@
+}
+
+extension PluginSetupWebViewController: WKNavigationDelegate {
+    func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
```
nit: maybe we can try using the async/await version `WKNavigationDelegate.webView(_:decidePolicyFor:) async`?

itsmeichigo replied:
Also updated in 45e46f67ba3ccfddaa50f0ec4fd4b5b3ee7bf5e4.

jaclync commented on WooCommerce/WooCommerceTests/Authentication/JetpackSetupWebViewModelTests.swift:
```diff
@@ -0,0 +1,85 @@
+import XCTest
+@testable import WooCommerce
+
+final class JetpackSetupWebViewModelTests: XCTestCase {
```
nice tests!

jaclync reviewed (approved):
LGTM 🚢 

When I tested the PR for the first time, I didn't reinstall the app and the app navigates to an unknown store (no data) after entering the site address (a JN site) then WP.com credentials without the Jetpack error screen 😅 However, I couldn't reproduce this after reinstalling the app, maybe it's some weird edge case.

itsmeichigo committed:
Make PluginSetupWebViewModel's title a constant

itsmeichigo committed:
Check view controller's dismissal state with `isBeingDismissedInAnyWay`

itsmeichigo committed:
Use async method for `decidePolicy`

itsmeichigo commented:
Thank you @jaclync for the review! I addressed the issues in the comments in the new commits.

> When I tested the PR for the first time, I didn't reinstall the app and the app navigates to an unknown store (no data) after entering the site address (a JN site) then WP.com credentials without the Jetpack error screen 

By "unknown store" do you mean the app navigates directly to the home screen? If so the site must still have Jetpack or for some reason, the state hadn't completed syncing, so you still got valid Jetpack when fetching the site info 🤔.

jaclync commented:
> By "unknown store" do you mean the app navigates directly to the home screen? If so the site must still have Jetpack or for some reason, the state hadn't completed syncing, so you still got valid Jetpack when fetching the site info 🤔.

Yea, the app navigates to the logged-in state after entering a JN site address then WP.com credentials without the Jetpack error screen. I'm quite sure it's the same JN site address that I just created, and later saw the Jetpack error screen after reinstalling the app. But I can't reproduce it anymore 😅

itsmeichigo commented:
I'm thinking maybe we should change the strategy to check the matched site from `ULMatcher` for Jetpack connection rather than keeping the reference of the site information. I'll look into that in another PR - merging this for now.
